File: gen.pl
#!/usr/bin/perl

use strict;
use warnings;
use File::Copy;
use JSON;
use Data::Dumper;
use autodie;

my $lang = shift(@ARGV) || 'pie';

#Read in template file
open my $fh, '<', 'template.json';

my $template;

#slurp in the file contents
{
  local $/ = undef;
  $template = JSON->new()->decode(<$fh>);
}

my $subdirs = $template->{'directories'};
my $mdDir = $template->{'mdDir'};
my $langDirs = $template->{'langDir'};
my $langPath = $langDirs . $lang;

if(!-e $langDirs){
  mkdir $langDirs;
}

if(!-e $langPath){
  mkdir $langPath;
}

copy($mdDir . 'Readme.md', $langPath . '/Readme.md');

for my $dir( @{ $subdirs }){
  my $path = $langPath . "/$dir";
  if(!-e $path){
    mkdir $path;
  }
  my $src = $mdDir . "/$dir" . '.md';
  my $dest = $path . "/Readme.md\n";
  copy($src, $dest);

  print "Created $dest" if -e $dest;
}

File: gen.js
var fs = require('fs');

//Read in template file
var contents = fs.readFileSync('template.json', 'utf-8');

//Convert to json
var template = JSON.parse(contents);

var lang = process.argv[2];

if(lang === undefined || lang === null){
  throw new Error('Must include a language name on cmd line');
}

var langDirs = template.langDir;

//Makes the directory that holds each languange if it
//has not been created
if(!fs.existsSync(langDirs)){
  fs.mkdirSync(langDirs);
}

var langPath = langDirs + lang;

var exists = fs.existsSync(langPath);

//May change how this part works
if(exists){
  console.log(langPath, 'exists, but continuing');
} else {
  fs.mkdir(langPath);
}

var mdDir = template.mdDir;

fs.writeFileSync(
  langPath + '/Readme.md',
  fs.readFileSync(mdDir + 'Readme.md')
);

var subdirs = template.directories;

//We can go async from here
subdirs.forEach(function(dir){
  var path = langPath +"/"+ dir;
  fs.exists(path, function(exists){
    if(!exists){
      fs.mkdirSync(path);
    }

    var src = mdDir + dir + '.md';
    var dest = path +'/Readme.md';
    copy(src, dest);
  });
});

//Asynchrous file copying
function copy(src, dest){
  if(!fs.existsSync(dest)){
    fs.readFile(src, function(err, data){
      if(err){
        console.log(err);
        return;
      }
      fs.writeFile(dest, data, function(err){
        if(err){
          console.log('Error writing to', dest);
        } else {
          console.log('Created ', dest);
        }
      });
    });
  }
}
